gh-99633: Add context manager support to `contextvars.Context` (#99634)
python/cpython: Doc/library/contextvars.rst

   behave in a similar fashion to :func:`threading.local` when values are
   assigned in different threads.

   Attempting to do either of the following raises a :exc:`RuntimeError`:

   * Entering an already entered context, including contexts entered in
     other threads.
   * Exiting from a context that is not the current context.

   After exiting a context, it can later be re-entered (from any thread).

   Any changes to :class:`ContextVar` values via the :meth:`ContextVar.set`
   method are recorded in the current context.  The :meth:`ContextVar.get`
   method returns the value associated with the current context.  Exiting a
   context effectively reverts any changes made to context variables while the
   context was entered (if needed, the values can be restored by re-entering the
   context).

   Context implements the :class:`collections.abc.Mapping` interface.

   .. versionadded:: 3.14
      Added support for the :term:`!context management protocol` (:pep:`343`).
      The value bound to the identifier given in the :keyword:`with` statement's
      :keyword:`!as` clause (if present) is the :class:`!Context` object itself.

>       Example:

Reviewer: The example should be out of the `.. versionchanged` section, ideally before (we should use `versionchanged` instead of `versionadded` since the latter is only when the object is being added, not when features are added).

   To get a copy of the current context use the
   :func:`~contextvars.copy_context` function.

>    Each thread has its own effective stack of :class:`~contextvars.Context`
>    objects.  The *current Context* is the Context object at the top of the
>    current thread's stack.  All Context objects in the stacks are considered to
>    be *entered*.  

Reviewer: ```suggestion
   Each thread has its own effective stack of :class:`Context` objects.
   The *current context* is the :class:`!Context` object at the top of
   the current thread's stack.  All :class:`!Context` objects in the
   stacks are considered to be *entered*.
```

>    *Entering* a Context, either by calling the
>    :meth:`Context.run` method or using the Context object as a :term:`context
>    manager`, pushes the Context onto the top of the current thread's stack,
>    making it the current Context.  *Exiting* from the current Context, either by
>    returning from the callback passed to :meth:`Context.run` or by exiting the
>    :keyword:`with` statement suite, pops the Context off of the top of the
>    stack, restoring the current Context to what it was before.

Reviewer: ```suggestion
   *Entering* a context, either by calling the :meth:`~Context.run` method
   or using the context object as a :term:`context manager`, pushes the
   context onto the top of the current thread's stack, making it the
   current context.

   *Exiting* from the current context, either by returning from the callback
   passed to :meth:`~Context.run` or by exiting the :keyword:`with` statement
   suite, pops the contextoff of the top of the stack, restoring the current
   context to what it was before.
```

>    Because each thread has its own Context stack, :class:`ContextVar` objects

Reviewer: ```suggestion
   Since each thread has its own context stack, :class:`ContextVar` objects
```

   behave in a similar fashion to :func:`threading.local` when values are
   assigned in different threads.

>    Attempting to do either of the following will raise a :exc:`RuntimeError`:

Reviewer: ```suggestion
   Attempting to do either of the following raises a :exc:`RuntimeError`:
```

>    * Entering an already entered Context.  (This includes Contexts entered in
>      other threads.)
>    * Exiting from a Context that is not the current Context.

Reviewer: ```suggestion
   * Entering an already entered context, including contexts entered in
     other threads.
   * Exiting from a context that is not the current context.
```

>    After exiting a Context, it can later be re-entered (from any thread).

Reviewer: ```suggestion
   After exiting a context, it can later be re-entered (from any thread).
```

>    Any changes to :class:`ContextVar` values via the :meth:`ContextVar.set`
>    method are recorded in the current Context.  The :meth:`ContextVar.get`
>    method returns the value associated with the current Context.  Thus, exiting
>    a Context effectively reverts any changes made to context variables while the
>    Context was entered.  (If desired, the values can be restored by re-entering
>    the Context.)

Reviewer: ```suggestion
   Any changes to :class:`ContextVar` values via the :meth:`ContextVar.set`
   method are recorded in the current context. The :meth:`ContextVar.get`
   method returns the value associated with the current context. Exiting
   a context effectively reverts any changes made to context variables while
   the context was entered (if needed, the values can be restored by re-entering
   the context).
```

   Context implements the :class:`collections.abc.Mapping` interface.

   .. versionadded:: 3.14

>       Context implements the :term:`!context management protocol` (:pep:`343`).

Reviewer: ```suggestion
      Added support for the :term:`context management protocol`.
```

Let's reduce the number of links.

Author: "Context management protocol" is not defined in the glossary, which is why I linked to the PEP instead.

I could do ``:term:`context manager` protocol`` instead, although that terminology doesn't exactly match "context management protocol" used in the PEP.

Or I could add "context management protocol" to the glossary.

WDYT?

Reviewer 2: Yep, let's do that.

   .. method:: __enter__(self)
               __exit__(self, exc_type, exc_value, exc_tb)

      These methods implement the :term:`!context management protocol`
      (:pep:`343`), making it possible to use a Context object as a
      :term:`context manager` with the :keyword:`with` statement.  The
      :meth:`__enter__` method enters the Context and returns *self*, so the
      value bound to the identifier given in the :keyword:`with` statement's
      :keyword:`!as` clause (if present) is the Context object itself.  The

>       :meth:`__exit__` method exits the Context.

Reviewer: ```suggestion
```

Other classes don't seem to document those kind of methods. Since they don't do much more than what you already explained, we can simplify the docs (but keep the examples).

      Example:

      .. testcode::

         import contextvars

         var = contextvars.ContextVar("var")
         var.set("initial")
         assert var.get() == "initial"

         # Copy the current Context and enter the copy.
         with contextvars.copy_context() as ctx:
             var.set("updated")
             assert var in ctx
             assert ctx[var] == "updated"
             assert var.get() == "updated"

         # Exited ctx, so the observed value of var has reverted.
         assert var.get() == "initial"
         # But the updated value is still recorded in ctx.
         assert ctx[var] == "updated"

         # Re-entering ctx restores the updated value of var.
         with ctx:

>              assert var.get() == "updated"

Reviewer: I'd suggest using the same formulation as in `run()` where we don't use `assert` but rather comments.

Author: Note the use of `.. testcode::`.  I want this example to be testable to prevent the implementation from drifting away from the documentation.  Would you prefer I use `print` and `.. testoutput::`?  I personally find `assert` easier to read.

Reviewer 2: Is that test code actually checked, though? I guess put differently: can you point me at the CPython CI that's actually executing this test code and checking it is as expected?

Reviewer: That would be the doctest CI job. This is a sphinx directive for running doctest: https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html#directive-testcode.

---

> Note the use of .. testcode::. I want this example to be testable to prevent the implementation from drifting away from the documentation. Would you prefer I use print and .. testoutput::? I personally find assert easier to read.

I think print() could be preferrable so that users, if they C/C the code, would also see the output directly. With a combination of `.. testoutput` that would be enough.

You can also change the rst for run() so that it also uses `testcode/testoutput`. Alternatively, you can use the `doctest` directive directly.

Author: I updated the `run` documentation to also use `testcode`/`testoutput`.

>       .. versionadded:: 3.14

Reviewer: ```suggestion
```
